### PR TITLE
crimson/os/seastore/cached_extent: add the "refresh" ability to lba mappings

### PR DESCRIPTION
**Pull request by xxhdx1985126**
So that we don't have to re-iterate through the whole lba tree to get a "parent-viewable" mapping in most cases.

This PR is a performance optimization for #57709, and is also the first step toward the lba mapping updating optimization discussed in https://github.com/ceph/ceph/pull/55449#discussion_r1479151298

Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.h [hunk 1]
```diff
@@ -176,6 +176,14 @@ class BtreeLBAMapping : public BtreeNodeMapping<laddr_t, paddr_t> {
     auto &p = static_cast<LBALeafNode&>(*parent);
     p.maybe_fix_mapping_pos(*this);
   }
+
+  LBAMappingRef refresh() final {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.h:180]: Worth to have a more explicit name, e.g.
`refresh_with_pending_parent()`

File: src/crimson/os/seastore/btree/fixed_kv_node.h [hunk 2]
```diff
@@ -125,6 +125,27 @@ struct FixedKVNode : ChildableCachedExtent {
     copy_dests.dests.erase(dest);
   }
 
+  FixedKVNodeRef find_dest(Transaction &t, node_key_t key) {
```
Comment (cyx1231st) [inline src/crimson/os/seastore/btree/fixed_kv_node.h:128]: `find_dest` -> `find_pending_version` ?

**Review by cyx1231st — COMMENTED** [review]
Generally looks reasonable!

After this we should have bidirectional pointers between parents and children, right?

The comments are mostly about better maintainence.

**Comment by xxhdx1985126** [comment]
> Generally looks reasonable!
> 
> After this we should have bidirectional pointers between parents and children, right?

Yeah, children can always find their parents, too.

File: src/crimson/os/seastore/btree/fixed_kv_node.h [hunk 3]
```diff
@@ -84,11 +84,47 @@ struct FixedKVNode : ChildableCachedExtent {
...
 
+  struct copy_dests_t : trans_spec_view_t {
+    std::set<FixedKVNodeRef, fixedkv_node_cmp_t> dests;
+    copy_dests_t(Transaction &t) : trans_spec_view_t{t.get_trans_id()} {}
+    ~copy_dests_t() {
+      LOG_PREFIX(~copy_dests_t);
+      SUBTRACE(seastore_fixedkv_tree, "copy_dests_t destroyed");
+    }
+  };
+
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/btree/fixed_kv_node.h:100]: Yes, that's right. Will add this comment into the code. Thanks.

File: src/crimson/os/seastore/cached_extent.h [hunk 4]
```diff
@@ -1277,7 +1277,7 @@ class ChildableCachedExtent : public CachedExtent {
   void reset_parent_tracker(parent_tracker_t *p = nullptr) {
     parent_tracker.reset(p);
   }
-  bool is_parent_valid() const {
+  bool is_parent_viewable() const {
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/cached_extent.h:1285]: Yes, `is_parent_valid()` is later defined as a method that checks whether the parent is literally valid.

[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/cached_extent.h:1285]: Still not sure, can you help point to its user?

Unlike `PhysicalNodeMapping::is_parent_valid()`, this(`ChildableCachedExtent`) looks to me that the implementation is simply checking whether the parent is valid.

So I'm not sure whether rename this is really helpful.

[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/cached_extent.h:1285]: `is_parent_viewable()` consider parents that are valid but not viewable by the transaction as false. I added a new `is_parent_valid()` in 642dbdcb2765b98a805889f7cefab00ee7da7067 which checks whether the parent is valid.

If the parent is valid but not viewable by the current transaction, we "refresh" the mapping; while, if the parent is not valid at all, we need to re-search the lba tree to get a valid mapping.

[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/cached_extent.h:1285]: The rename is just to make the name of the method more accurate.

[the remark this answers is not in the document]
Comment (cyx1231st) [reply src/crimson/os/seastore/cached_extent.h:1285]: Yeah, I think I'm fully understanding and also agree with the renaming `pin->is_parent_viewable()`.

Only not sure about `ChildableCachedExtent::is_parent_valie -> viewable()` from its implementation:
https://github.com/ceph/ceph/blob/642dbdcb2765b98a805889f7cefab00ee7da7067/src/crimson/os/seastore/cached_extent.h#L1285-L1287

[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/cached_extent.h:1285]: Ah, sorry, this is a mistake, `ChildableCachedExtent` shouldn't be touched.

**Comment by xxhdx1985126** [comment]
@cyx1231st All addressed, please take a look again, thanks:-)

File: src/crimson/os/seastore/btree/fixed_kv_node.h [hunk 5]
```diff
@@ -84,11 +84,55 @@ struct FixedKVNode : ChildableCachedExtent {
...
+    auto tid = t.get_trans_id();
+    auto iter = copy_dests_by_trans.lower_bound(
+      tid, trans_spec_view_t::cmp_t());
+    if (iter == copy_dests_by_trans.end() ||
+	iter->pending_for_transaction != tid) {
+      iter = copy_dests_by_trans.insert_before(
+	iter, t.add_transactional_view<copy_dests_t>(t));
+    }
+    auto &copy_dests = static_cast<copy_dests_t&>(*iter);
+    copy_dests.dests_by_key.insert(dest);
```
Comment (cyx1231st) [inline src/crimson/os/seastore/btree/fixed_kv_node.h:123]: Is it safer to assert that the insertion must be successful (insertion must took place) ?

**Review by cyx1231st — COMMENTED** [review]
LGTM except one comment.

File: src/crimson/os/seastore/btree/fixed_kv_node.h (hunk 5 above)
Comment (xxhdx1985126) [reply src/crimson/os/seastore/btree/fixed_kv_node.h:123]: Um, actually, there seems to be cases in which duplicated insertions are inevitable, for example:

1. stable LBANode `L1` is split into two pending `L2` and `L3`
2. after a while, the same transaction merges `L2` and `L3` into `L4`
3. when pusing copy sources, the transaction may try to add `L4` into `L1`'s `dests_by_key` twice.

Comment (cyx1231st) [reply src/crimson/os/seastore/btree/fixed_kv_node.h:123]: Thanks for the scenario!

I see, how about implementing an assert that if `dest` cannot be inserted, the conflicting extent must be the `dest`.

Is that reasonable?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/btree/fixed_kv_node.h:123]: yes, will do

Comment (xxhdx1985126) [reply src/crimson/os/seastore/btree/fixed_kv_node.h]: Done:-)

**Comment by cyx1231st** [comment]
make check failed by:
> ERROR 2024-07-19 10:55:48,460 seastar - width/precision is not integer

**Comment by cyx1231st** [comment]
jenkins test make check

**Review by cyx1231st — APPROVED** [review]

File: src/crimson/os/seastore/cached_extent.h [hunk 6]
```diff
@@ -330,7 +333,7 @@ class CachedExtent
    * Called when this extent is rewriting another one.
    *
    */
-  virtual void on_rewrite(CachedExtent &, extent_len_t) = 0;
+  virtual void on_rewrite(CachedExtent &, extent_len_t, Transaction &) = 0;
```
[the remark this answers is not in the document]
Comment (xxhdx1985126) [reply src/crimson/os/seastore/cached_extent.h]: Sorry, I missed this comment, just addressed it:-)

**Comment by cyx1231st** [comment]
@xxhdx1985126 I was preparing a testing branch, need to know what was changed^

**Comment by xxhdx1985126** [comment]
> @xxhdx1985126 I was preparing a testing branch, need to know what was changed^
> 
> https://shaman.ceph.com/builds/ceph/wip-yingxin-testing-2024-07-22-1041/

Only the parameter order of "CachedExtent::on_rewrite()", no need to repush the testing branch:-)

**Comment by cyx1231st** [comment]
https://pulpito.ceph.com/yingxin-2024-07-22_05:35:39-crimson-rados-wip-yingxin-testing-2024-07-22-1041-distro-default-smithi
* -7812564/7812565: https://tracker.ceph.com/issues/63844

**Comment by cyx1231st** [comment]
jenkins test api

**Comment by cyx1231st** [comment]
jenkins test api

**Pull request merged**